R/dsl.R

```diff
@@ -132,11 +133,69 @@ add_package_checks <- function(...,
       add_code_step(covr::codecov(quiet = FALSE))
   }
 }
+#' @description
+#' `do_pkgdown_site()` builds and optionally deploys a pkgdown site.
+#'
+#'
+#' @inheritParams step_build_pkgdown
+#' @inheritParams step_setup_push_deploy
+#' @inheritParams step_do_push_deploy
+#' @param build_only Build the pkgdown site but do not deploy it.
+#' @param deploy Checks if env variable `id_rsa` is set in Travis using [ci_can_push()]. If missing,
+#'   deployment is not possible.
+#' @rdname DSL
+#' @export
+#' @importFrom magrittr %>%
+do_pkgdown_site <- function(...,
+  build_only = FALSE,
+  deploy = ci_can_push(),
+  orphan = FALSE,
+  checkout = TRUE,
+  repos = getOption("repos"),
+  path = ".", branch = NULL,
+  remote_url = NULL,
+  commit_message = NULL, commit_paths = ".") {
+
+  #' @description
+  #' 1. [step_install_deps()] in the `"install"` stage, using the
+  #'    `repos` argument.
+  get_stage("install") %>%
+    add_step(
+      step_install_deps(repos = repos)
+    )
+
+  #' 1. [step_setup_ssh()] in the `"before_deploy"` to setup the upcoming deployment.
+  get_stage("before_deploy") %>%
+    add_step(
+      step_setup_ssh()
+    )
+
+  #' 1. [step_build_pkgdown()] in the `"deploy"` stage
+  get_stage("deploy") %>%
+    add_step(
+      step_build_pkgdown(...)
+    )
+
+  if (isTRUE(build_only) | !deploy) {
+    ci_cat_with_color("`build_only = TRUE` was set, skipping deployment")
+  } else {
+    #' 1. [step_setup_push_deploy()] in the `"deploy"` stage.
+    #' 1. [step_do_push_deploy()] in the `"deploy"` stage. By default, the deploy is done to the gh-pages branch.
+    #'
+    if (ci_can_push()) {
+      get_stage("deploy") %>%
+        add_step(step_setup_push_deploy(path = path, branch = branch,
```

Comment:
This must be added before adding the `step_build_pkgdown()` .

Reply:
After adding `deploy = ci_can_push()` we do not need it at all in the function, right?

Why would we need it before `step_build_pkgdown()`? If we only want to build but not deploy (e.g. if there is no `id_rsa` set), the building would be skipped by adding it.

Currently

- building + deployment by default if `id_rsa` exists
- building only if `build_only = TRUE` or `id_rsa` missing

Reply:
The sequence needs to be:

- setup deploy, build, do deploy if we can push
- build if we can't push

Otherwise `gh-pages` is always orphaned.

Maybe I don't fully understand the question?

Reply:
The question is: Why should we setup deployment if we do not deploy? I was not aware that we need `step_setup_push_deploy()` to run `step_build_pkgdown()` if we actually do not want to deploy?

Maybe we are both confused now.
Anyhow, I simply changed the order :smile: Should be fine now?

Reply:
We setup only if we deploy, we don't need it for building the pkgdown.